### PR TITLE
Lease domain

### PR DESCRIPTION
**Pull request by SimonRichardson**
The lease state was in the worker package, this was prior to the domain 
package. This moves that over to align it with all the other pieces of domain
code that we've actively been working on.

This also allows us to separate the store from the validation logic
which defines the service type.

As a driveby, I moved the manifold back to the worker package.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju enable-ha
$ juju deploy ubuntu -n 3
```

### CODE REVIEW

File: domain/lease/state/state_test.go [hunk 1]
```diff
@@ -1,39 +1,40 @@
...
 
-func (s *storeSuite) SetUpTest(c *gc.C) {
+func (s *stateSuite) SetUpTest(c *gc.C) {
 	s.ControllerSuite.SetUpTest(c)
 
-	s.store = lease.NewStore(lease.StoreConfig{
-		TxnRunner: s.TxnRunner(),
-		Logger:    lease.StubLogger{},
-	})
+	s.store = state.NewState(domain.NewTxnRunnerFactory(s.getWatchableDB))
```
Comment (manadart) [inline domain/lease/state/state_test.go:34]: You should be able to do `state.NewState(c.TxnRunner)`

File: domain/lease/state/state_test.go [hunk 2]
```diff
@@ -251,6 +253,19 @@ func (s *storeSuite) TestLeaseOperationCancellation(c *gc.C) {
...
+
+func (s *stateSuite) getWatchableDB() (changestream.WatchableDB, error) {
+	return &stubWatchableDB{TxnRunner: s.TxnRunner()}, nil
+}
+
+func (s *stateSuite) getWatchableDBForNameSpace(_ string) (changestream.WatchableDB, error) {
+	return &stubWatchableDB{TxnRunner: s.TxnRunner()}, nil
+}
+
+type stubWatchableDB struct {
```
Comment (manadart) [inline domain/lease/state/state_test.go:268]: Which means you don't need this.

File: worker/lease/manifold.go [hunk 3]
```diff
@@ -155,11 +144,13 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
...
-func NewWorker(config lease.ManagerConfig) (worker.Worker, error) {
-	return lease.NewManager(config)
+func NewWorker(config ManagerConfig) (worker.Worker, error) {
+	return NewManager(config)
 }
 
 // NewStore returns a new lease store based on the input config.
-func NewStore(config lease.StoreConfig) *lease.Store {
-	return lease.NewStore(config)
+func NewStore(runner coredatabase.TxnRunner) lease.Store {
```
Comment (manadart) [inline worker/lease/manifold.go:152]: I would do this:
```go
func NewStore(getter coredatabase.DBGetter) lease.Store {
    return service.NewService(domain.NewTxnRunnerFactoryForNamespace(getter.GetDB, coredatabase.ControllerNS))
}
```

Then further up you don't get the DB; it's just:
```go
store := s.config.NewStore(dbGetter)
```

**Review by manadart — COMMENTED** [review]
Some suggestions.

File: domain/lease/state/state_test.go (hunk 1 above)
Comment (SimonRichardson) [reply domain/lease/state/state_test.go:34]: Nearly. We had to add `testing.TxnRunnerFactory` to get around the function return mismatch.

```go
s.store = state.NewState(testing.TxnRunnerFactory(s.TxnRunner()))
```

File: worker/lease/manifold.go (hunk 3 above)
Comment (SimonRichardson) [reply worker/lease/manifold.go:152]: Yeah, that is better.

**Review by manadart — APPROVED** [review]
Thanks for the changes. Looks good.

**Pull request merged**